text/0000-float-semantics.md

```diff
@@ -0,0 +1,268 @@
+
+The non-deterministic choice happens when the operation is executed; i.e., the result of a NaN-producing floating point operation is a stable bit pattern (looking at these bits multiple times will yield consistent results), but running the same operation twice with the same inputs can produce different results.
+
+Unless noted otherwise, the same rules also apply to NaNs returned by other library functions (e.g. `min`, `minimum`, `max`, `maximum`); other aspects of their semantics and which IEEE 754-2008 operation they correspond to are documented with the respective functions.
+
+### `const` semantics
+
+Const-evaluation must necessarily be deterministic to ensure soundness of the type system.
+`const` use of floating points does not make any guarantees beyond that:
+when a floating-point operation produces a NaN result, the resulting NaN bit pattern is *some* deterministic function of the operation's inputs that satisfies the constraints placed on run-time floating point semantics.
+However, the exact function is not specified, and it is allowed to change across targets and Rust versions, and even with compiler flags.
+In particular, there is no guarantee that the choice made in const evaluation is consistent with the choice made at runtime.
+That is, the following assertion is allowed to fail (and in fact, it [fails on current versions of Rust](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a594d2975c29b1c7fa457a4ec4ae4b87)):
+```rust
+#![allow(non_snake_case)]
+use std::hint::black_box;
+
+const C: f32 = 0.0 / 0.0;
+
+fn main() {
+    let c: f32 = 0.0 / black_box(0.0);
+    assert_eq!(C.to_bits(), c.to_bits());
+}
+```
+
+This means that evaluating the same `const fn` on the same arguments can produce different results at compile-time and run-time.
+However, note that these functions are already non-deterministic: even evaluating the same function with the same arguments twice at runtime can [and does](https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=50b5a549fa1fe259cea5ad138066ccf0) produce different results!
+
+This resolves the last open question blocking floating-point operations in `const fn`.
+When the RFC is accepted, the `const_fn_floating_point_arithmetic` feature gate and the `const fn to_bits` methods can be stabilized.
+
+### Assumptions about floating-point environment
+
+This RFC is primarily concerned with the guarantee Rust provides to its users.
+How exactly those guarantees are achieved is an implementation detail, and not observable when writing pure Rust code.
+However, when mixing Rust with inline assembly, those details *do* become observable.
+To ensure that Rust can provide the above guarantees to user code, it is UB for inline assembly to alter the behavior of floating-point operations in any way: when leaving the inline assembly block, the floating-point environment must be in exactly the same state as when the inline assembly block was entered.
```

Comment:
Possibly, but it would probably have to be fairly specific and would rely on LLVM promising more things. We presently have an example causing a segfault where the only unsafe thing is setting the rounding mode: https://github.com/rust-lang/unsafe-code-guidelines/issues/471#issuecomment-1774261953

Reply:
I don't think we can do better than "UB" here. When an operation does not produce the result that the spec says it produces, there's no limit to what happens. People can write `if 1+1 != 2 { unreachable_unchecked() }`, and they can do the same with float operations, so changing the rounding mode can lead to arbitrary misbehavior including UB.

Reply:
But this is the spec we're writing now, isn't it, so if the spec was weaker they couldn't rely on it? Can we weaken it to say something like "in the absence of code which modifies these flags, floats behave according to IEEE 754-2008" and "if you change any of these flags, optimizations will cause your code to behave nondeterministically"?

Reply:
What's the difference between "behave nondeterministically" and UB?

Reply:
Sorry, what I meant was that the _result of floating point operations_ would be nondeterministic. But we can still put bounds around what they would do; e.g. they would always return a value if you haven't set any exception flags.

Reply:
That's strengthening the spec, not weakening it. In any case, we cannot, as the aforementioned example shows. LLVM will happily optimize away (integer!) array index bounds checks if it can prove that they are never needed under default rounding.

Reply:
> we can still put bounds around what they would do

That's exactly the point, we cannot. (a) there is the [example](https://github.com/rust-lang/unsafe-code-guidelines/issues/471#issuecomment-1774261953), and (b) the compile ris allowed to move floating-point operations around, so even if you wrote the FP operation outside the block where the flags are different, it might end up executing inside the block.

I don't know any alternative to declaring this UB.